Add skeletons of the inter-era translations needed by consensus

mrBliss opened the pull request:
See the individual commits for more details.

mrBliss committed:
Export TranslationContext

mrBliss committed:
Move back orphan Eq instances to their types

While I agree with the idea of this change, propagating it to consensus is more
complex (e.g., composability of `Eq` instances in the HFC). I'd like to make a
similar change in consensus, but later.

mrBliss committed:
Add translateEraMaybe

mrBliss committed:
Add a placeholder instance of Compactible (Value era)

`ShelleyBased` requires this instance, so without it consensus can't add the
Mary era.

Obviously, this is not a proper instance.

mrBliss committed:
Add skeletons of the inter-era translations needed by consensus

See the comments in the modules for more details. The new instances should be
implemented by the ledger team and kept up-to-date as the Allegra and Mary
ledger evolve.

mrBliss commented on shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs:
```diff
@@ -0,0 +1,110 @@
+type instance TranslationContext (AllegraEra c) = ()
+
+instance Crypto c => TranslateEra (AllegraEra c) ShelleyState where
+  translateEra _ = undefined
```
All the `undefined`s should be implemented, of course.

nc6 replied:
You don't say!

mrBliss commented on shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs:
```diff
@@ -0,0 +1,110 @@
+              $ genesis,
+          sgInitialFunds = Map.mapKeys (translateEra' ctxt) $ sgInitialFunds genesis,
+          sgStaking = translateEra' ctxt $ sgStaking genesis
+        }
```
This is an example of how you _might_ implement a translation.

mrBliss commented on shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs:
```diff
@@ -0,0 +1,110 @@
+--------------------------------------------------------------------------------
+
+instance Crypto c => TranslateEra (AllegraEra c) HashHeader where
+  translateEra _ (HashHeader hash) = return $ HashHeader $ Hash.castHash hash
```
Another example

mrBliss commented on shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs:
```diff
@@ -0,0 +1,110 @@
+  translateEra _ = undefined
+
+instance Crypto c => TranslateEra (AllegraEra c) Addr where
+  translateEra _ = undefined
```
The instance for `ShelleyGenesis` needed these auxiliary instances. When CAD-2000 is implemented, some of them can go (`KeyHash 'Genesis`, `GenDelegPair`).

mrBliss commented on shelley-ma/impl/src/Cardano/Ledger/Allegra/Translation.hs:
```diff
@@ -0,0 +1,110 @@
+--------------------------------------------------------------------------------
+-- The following instances are also needed by consensus, but only temporarily.
+-- These types should not be parameterised by the era but by the crypto. When
+-- this has changed (CAD-2000), the instances below can be removed.
```
@JaredCorduan You could say that CAD-2000 is blocking this PR and this PR is blocking the [next step of the integration in consensus](https://github.com/input-output-hk/ouroboros-network/pull/2679) (i.e., making Allegra and Mary actual different eras, not just Shelley clones) :slightly_smiling_face:

JaredCorduan replied:
thanks for the heads up @mrBliss , I'm working on CAD-2000 now.

JaredCorduan replied:
I was hoping to finish this up today, but I'm still not there. my branch is `jc/remove-era-from-protocol`, it's down to the test code. here's hoping tomorrow brings it to an end :crossed_fingers:

mrBliss replied:
Great! I'll first rebase this PR on master and then have a look at your branch (and maybe start rebasing on that branch).

mrBliss replied:
Done: rebased this PR on master. I also created `mrBliss/era-translations2`, which is this branch rebased on `jc/remove-era-from-protocol`. That branch is already a bit out of date w.r.t. `master`, though.

@JaredCorduan Your branch allowed me to get rid of a bunch of `TranslateEra` instances, as expected :+1:.

I'll now test this out with my draft PR in consensus.

mrBliss replied:
@JaredCorduan Propagating `jc/remove-era-from-protocol` feels so right :relaxed:.

So far, I have encountered one problem: consensus needs to pass the `PParams` to `chainChecks`. Before your changes, we could extract them from the `LedgerView`, but that's no longer possible after your changes. Consensus doesn't have anything else in scope where it could get the `PParams` from. Potential solution: create a record with the fields needed by `chainChecks` (`_maxBHSize`, `_maxBBSize`, `_protocolVersion`), let `chainChecks` that record as an argument and store the record in the `LedgerView`.

mrBliss replied:
@JaredCorduan Another thing that I noticed: we have the following class in consensus:
```haskell
class ( Crypto c
      , SL.DSignable    c (OCertSignable c)
   -- , SL.DSignable    c (Hash era (TxBody era)) TODO
      , SL.KESignable   c (BHBody c)
      , SL.VRFSignable  c Seed
      ) => TPraosCrypto c
```
It would be a lot more convenient if that `era` wasn't there :slightly_smiling_face:.

Maybe we can do something similar to what you did for the block body? I.e.:

```haskell
newtype HashBBody crypto = UnsafeHashBBody {unHashBody :: (Hash crypto EraIndependentBodyHash)}
```

JaredCorduan replied:
ok, thank you for all the clear feedback @mrBliss ! I'm glad to hear that this is mostly what you expected :)

mrBliss replied:
Alternatively, adding `SL.DSignable (EraCrypto era) (Hash (EraCrypto era) (TxBody era))` to `ShelleyBased` would also be convenient.

mrBliss committed:
Add Foldable/Traversable instances for StrictMaybe

mrBliss committed:
Swap order of WitVKey type arguments

Put `era` last, like in the other types. Otherwise we can't use it with
`TranslateEra` (without jumping through extra hoops), as it expects the `era` to
be the last type argument.

mrBliss committed:
Add missing bang to wvKeyHash

Either add a bang or, if it *should* be lazy, update `AllowThunksIn`.

mrBliss committed:
Update the documentation of TranslateEra

mrBliss commented:
Rebased.

This includes #1922 until merged.

@nc6 All commits in this PR, other than the *first* (2e15914) and *last* (2c4293a) can be merged.

I rely on the last one for the integration in consensus, but these instances haven't yet been implemented. How do you feel about merging incomplete code, containing a bunch of `undefined`s (we can change that to `error` if preferred)? 

My thinking is that the propagation of the actual Allegra and Mary eras (in consensus/node they currently are just aliases of `ShelleyEra`) to consensus/node/db-sync/wallet doesn't have to wait until all translations have been implemented. The translations won't be needed until we want to actually hard fork to the these eras. In the meantime, by having the real eras, the downstream repos can start preparing with the right types (instead of just aliases to Shelley). This means we can have more parallelisation in the process.

Nevertheless, it would be good to implement these translations soon, so that they can be updated incrementally when changes are made to Allegra/Mary, instead of having to write them from scratch at the end and then realising some things don't work about.

nc6 commented:
I'm fine having `undefined` so long as it's confined to the `shelley-ma` package (which is under construction, after all) and is needed for integration.

mrBliss commented:
> I'm fine having `undefined` so long as it's confined to the `shelley-ma` package (which is under construction, after all) and is needed for integration.

Okay, great :+1:. I'll rebase this branch as soon as #1922 is merged.

mrBliss committed:
Add missing ghc-options to cardano-ledger-shelley-ma

mrBliss commented:
Rebased! This is ready for review/merging. After this PR, I'll rebase #1923, and when that PR is merged, we can finally update the ledger dependency in consensus.

I have replaced the `undefined`s with `error`s with a slightly cleaner error message. @nc6 @JaredCorduan Should I should create an issue that these error messages can refer to?

nc6 commented on shelley-ma/impl/cardano-ledger-shelley-ma.cabal:
```diff
@@ -43,4 +45,11 @@ library
+    -Wincomplete-record-updates
+    -Wincomplete-uni-patterns
+    -Wredundant-constraints
+    -Wpartial-fields
```
:+1:
